File: src/ProtoWorkerMessageProcessor.java
package org.jetbrains.jps.bazel;

import com.google.devtools.build.lib.worker.WorkerProtocol.WorkRequest;
import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public final class ProtoWorkerMessageProcessor implements WorkRequestHandler.WorkerMessageProcessor {
  /**
   * This worker's stdin.
   */
  private final InputStream stdin;

  /**
   * This worker's stdout. Only {@link WorkRequest}s should be written here.
   */
  private final OutputStream stdout;

  /**
   * Constructs a {@link WorkRequestHandler} that reads and writes Protocol Buffers.
   */
  public ProtoWorkerMessageProcessor(InputStream stdin, OutputStream stdout) {
    this.stdin = stdin;
    this.stdout = stdout;
  }

  @Override
  public WorkRequest readWorkRequest() throws IOException {
    return WorkRequest.parseDelimitedFrom(stdin);
  }

  @Override
  public void writeWorkResponse(WorkResponse workResponse) throws IOException {
    try {
      workResponse.writeDelimitedTo(stdout);
    }
    finally {
      stdout.flush();
    }
  }

  @Override
  public void close() {
  }
}

File: src/ProtoWorkerMessageProcessor.kt
package org.jetbrains.bazel.jvm

import com.google.devtools.build.lib.worker.WorkerProtocol.WorkRequest
import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse
import java.io.InputStream
import java.io.OutputStream

class ProtoWorkerMessageProcessor(
  private val input: InputStream,
  private val output: OutputStream
) : WorkerMessageProcessor {
  override fun readWorkRequest(): WorkRequest? = WorkRequest.parseDelimitedFrom(input)

  override fun writeWorkResponse(workResponse: WorkResponse) {
    try {
      workResponse.writeDelimitedTo(output)
    }
    finally {
      output.flush()
    }
  }

  override fun close() {
  }
}